packages/grid/_modules_/grid/components/panel/filterPanel/GridFilterInputMultipleSingleSelect.tsx

```diff
@@ -0,0 +1,153 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import Autocomplete, { AutocompleteProps, createFilterOptions } from '@mui/material/Autocomplete';
+import Chip from '@mui/material/Chip';
+
+import TextField from '@mui/material/TextField';
+import { unstable_useId as useId } from '@mui/material/utils';
+import { GridFilterItem } from '../../../models/gridFilterItem';
+
+export type GridFilterInputMultipleSingleSelectProps = {
+  item: GridFilterItem;
+  applyValue: (value: GridFilterItem) => void;
+  // Is any because if typed as GridApiRef a dep cycle occurs. Same happens if ApiContext is used.
+  apiRef: any;
+  focusElementRef?: React.Ref<any>;
+  type?: 'singleSelect';
+} & Omit<AutocompleteProps<any[], true, false, true>, 'options' | 'renderInput'>;
+
+const getSingleSelectOptionFormatter =
+  ({ valueFormatter, field }, api) =>
+  (option) => {
+    if (typeof option === 'object') {
+      return option.label;
+    }
+    return valueFormatter && option !== '' ? valueFormatter({ value: option, field, api }) : option;
+  };
+
+const getValueFromOption = (option) => (typeof option === 'object' ? option.value : option);
+
+const isOptionEqualToValue = (option, value) =>
+  getValueFromOption(option) === getValueFromOption(value);
+
+const filter = createFilterOptions<any>();
+
+function GridFilterInputMultipleSingleSelect(props: GridFilterInputMultipleSingleSelectProps) {
+  const { item, applyValue, type, apiRef, focusElementRef, ...other } = props;
+  const [filterValueState, setFilterValueState] = React.useState(item.value || []);
+  const id = useId();
+
+  const currentColumn = item.columnField ? apiRef.current.getColumn(item.columnField) : null;
+  const currentValueOptions = React.useMemo(() => {
+    return typeof currentColumn.valueOptions === 'function'
+      ? currentColumn.valueOptions({ field: currentColumn.field })
+      : currentColumn.valueOptions;
+  }, [currentColumn]);
+
+  const filterValueOptionFormatter = getSingleSelectOptionFormatter(
+    apiRef.current.getColumn(item.columnField),
+    apiRef.current,
+  );
+
+  React.useEffect(() => {
+    let itemValue;
+    if (!Array.isArray(item.value)) {
+      itemValue = [];
+    } else if (currentValueOptions !== undefined) {
+      // sanitize if valueOptions are provided
+      const parsedCurrentValueOptions = currentValueOptions.map(getValueFromOption);
+      const itemValueIndexes = item.value.filter((element) => {
+        const formatedElement = getValueFromOption(element);
+        const index = parsedCurrentValueOptions.findIndex(
+          (formatedOption) => formatedOption === formatedElement,
+        );
+
+        return index;
+      });
+      itemValue = itemValueIndexes
+        .filter((index) => index >= 0)
+        .map((index) => currentValueOptions[index]);
+
+      if (itemValue.length !== item.value.length) {
+        // remove filtered values
+        applyValue({ ...item, value: parsedCurrentValueOptions });
```

Comment:
This is causing an infinite loop. Maybe we're missing a test case covering the selection of items to filter with a click, not only changing props.

Reply:
Yes, tests were made only on the operator "is", so I added similar ones on "isAnyOf" to verify if the values are cleaned when changing the column.

I also refactorized how value is sanitized. The AutoComplete requires to have coherent values with options, so I sanitize the value with a useMemo. Such that the sanitization is performed before rendering the AutoComplete component.

If sanitization removes some elements, the DataGrid state is updated by a dedicated hook.

I spend too much time on this little piece of code. I'm taking a break on it, and I come back tomorrow to check on the deployment if it is working as expected, or if bugs are remaining

Reply:
I played with filtering, no trivial bug was found this time